File: WNDS_Chapter_6/wnds_chapter_6a.R
# Defining and Visualizing Simple Networks (R)

# begin by installing necessary package igraph 

# load package into the workspace for this program
library(igraph)  # network/graph functions

# -------------------------
# star (undirected network)
# -------------------------
# define graph object for undirected star network
# using links for pairs of nodes by number
star <- graph.formula(1-2, 1-3, 1-4, 1-5, 1-6, 1-7)
# examine the graph object
print(str(star))

# name the nodes (vertices)
V(star)$name <- c("Amy", "Bob", "Dee", "Joe", "Lea", "Max", "Tia")

# examine the degree of each node
print(degree(star))

# determine the total number of links for the star (n-1)
print(sum(degree(star))/2)

# create an adjacency matrix object for the star network
star_mat <- get.adjacency(star)
print(star_mat)  # undirected networks are symmetric

# plot the network/graph to the console
plot(star, vertex.size = 20, vertex.color = "yellow")
# plot the network/graph and degree distribution to an external file
pdf(file = "fig_star_network.pdf", width = 5.5, height = 5.5)
plot(star, vertex.size = 25, vertex.color = "yellow", edge.color = "black")
hist(degree(star), col = "darkblue", 
    xlab = "Node Degree", xlim = c(0,8), main = "",
    breaks = c(0,1,2,3,4,5,6,7,8))
dev.off()

# ---------------------------
# circle (undirected network)
# ---------------------------
# define graph object for undirected circle network
# using links for pairs of nodes by number
circle <- graph.formula(1-2, 1-7, 2-3, 3-4, 4-5, 5-6, 6-7)
# examine the graph object
print(str(circle))

# name the nodes (vertices)
V(circle)$name <- c("Abe", "Bea", "Dag", "Eve", "Jim", "Kat", "Rob")

# examine the degree of each node
print(degree(circle))

# determine the total number of links 
print(sum(degree(circle))/2)

# create an adjacency matrix object for the circle network
circle_mat <- get.adjacency(circle)
print(circle_mat)  # undirected networks are symmetric

# plot the network/graph to the console
plot(circle, vertex.size = 20, vertex.color = "yellow")
# plot the network/graph and degree distribution to an external file
pdf(file = "fig_circle_network.pdf", width = 5.5, height = 5.5)
plot(circle, vertex.size = 25, vertex.color = "yellow", edge.color = "black")
hist(degree(circle), col = "darkblue", 
    xlab = "Node Degree", xlim = c(0,8), main = "",
    breaks = c(0,1,2,3,4,5,6,7,8))
dev.off()

# -------------------------
# line (undirected network)
# -------------------------
# define graph object for undirected line network
# using links for pairs of nodes by number
line <- graph.formula(1-2, 1-3, 2-4, 3-5, 4-6, 5-7)
# examine the graph object
print(str(line))

# name the nodes (vertices)
V(line)$name <- c("Ali", "Ben", "Ela", "Ian", "Mya", "Roy", "Zoe")

# examine the degree of each node
print(degree(line))

# determine the total number of links (n-1)
print(sum(degree(line))/2)

# create an adjacency matrix object for the line network
line_mat <- get.adjacency(line)
print(line_mat)  # undirected networks are symmetric

# plot the network/graph to the console
plot(line, vertex.size = 20, vertex.color = "yellow")
# plot the network/graph and degree distribution to an external file
pdf(file = "fig_line_network.pdf", width = 5.5, height = 5.5)
plot(line, vertex.size = 25, vertex.color = "yellow", edge.color = "black")
hist(degree(line), col = "darkblue", 
    xlab = "Node Degree", xlim = c(0,8), main = "",
    breaks = c(0,1,2,3,4,5,6,7,8))
dev.off()

# ---------------------------
# clique (undirected network)
# ---------------------------
# define graph object for undirected clique
# using links for pairs of nodes by number
a_clique <- graph.formula(1-2, 1-3, 1-4, 1-5, 1-6, 1-7,
                               2-3, 2-4, 2-5, 2-6, 2-7,
                                    3-4, 3-5, 3-6, 3-7,
                                         4-5, 4-6, 4-7,
                                              5-6, 5-7,
                                                   6-7)
# examine the graph object
print(str(a_clique))

# name the nodes (vertices)
V(a_clique)$name <- c("Ada", "Ala", "Ami", "Ana", "Ann", "Ara", "Ava")

# examine the degree of each node
print(degree(a_clique))

# determine the total number of links n(n-1)/2
print(sum(degree(a_clique))/2)

# create an adjacency matrix object for the clique
# this is a matrix of ones off-diagonal... fully connected
a_clique_mat <- get.adjacency(a_clique)
print(a_clique_mat)  # undirected networks are symmetric

# plot the network/graph to the console
plot(a_clique, vertex.size = 20, vertex.color = "yellow", 
    edge.color = "black", layout = layout.circle)
# plot the network/graph and degree distribution to an external file
pdf(file = "fig_clique_network.pdf", width = 5.5, height = 5.5)
plot(a_clique, vertex.size = 25, vertex.color = "yellow", 
    edge.color = "black", layout = layout.circle)
hist(degree(a_clique), col = "darkblue", 
    xlab = "Node Degree", xlim = c(0,8), main = "",
    breaks = c(0,1,2,3,4,5,6,7,8))
dev.off()

# -------------------------------
# tree (directed network/digraph)
# -------------------------------
# define graph object for undirected tree network
# using links for pairs of nodes by number
tree <- graph.formula(1-+2, 1-+3, 2-+4, 2-+5, 3-+6, 6-+7, 6-+8)
# examine the graph object
print(str(tree))

# name the nodes (vertices)
V(tree)$name <- c("Art", "Bev", "Dan", "Fay", "Lee", "Mia", "Sal", "Van")

# examine the degree of each node
print(degree(tree))

# create an adjacency matrix object for the tree network
tree_mat <- get.adjacency(tree)
print(tree_mat)  # directed networks are not symmetric

# plot the network/graph to the console 
plot(tree, vertex.size = 20, vertex.color = "yellow", edge.color = "black")

# plot the network/graph and degree distribution to an external file
# examine alternative layouts for plotting the tree 
pdf(file = "fig_tree_network_four_ways.pdf", width = 5.5, height = 5.5)
par(mfrow = c(1,1))  # four plots on one page
plot(tree, vertex.size = 25, vertex.color = "yellow", 
    layout = layout.fruchterman.reingold, edge.color = "black")
title("Fruchterman-Reingold Layout")   
plot(tree, vertex.size = 25, vertex.color = "yellow", 
    layout = layout.kamada.kawai, edge.color = "black")  
title("Kamada-Kawai Layout")    
plot(tree, vertex.size = 25, vertex.color = "yellow", 
    layout = layout.circle, edge.color = "black")
title("Circle Layout")     
plot(tree, vertex.size = 25, vertex.color = "yellow", 
    layout = layout.reingold.tilford, edge.color = "black")    
title("Reingold-Tilford Layout")       
hist(degree(tree), col = "darkblue", 
    xlab = "Node Degree", xlim = c(0,8), main = "",
    breaks = c(0,1,2,3,4,5,6,7,8))
dev.off()

File: WNDS_Chapter_6/wnds_chapter_6a.py
# Defining and Visualizing Simple Networks (Python)

# prepare for Python version 3x features and functions
from __future__ import division, print_function

# load package into the workspace for this program
import networkx as nx
import matplotlib.pyplot as plt  # 2D plotting
import numpy as np

# -------------------------
# star (undirected network)
# -------------------------
# define graph object for undirected star network
# adding one link at a time for pairs of nodes
star = nx.Graph()
star.add_edge('Amy', 'Bob')
star.add_edge('Amy', 'Dee')
star.add_edge('Amy', 'Joe')
star.add_edge('Amy', 'Lea')
star.add_edge('Amy', 'Max')
star.add_edge('Amy', 'Tia')

# examine the degree of each node
print(nx.degree(star))

# plot the star network and degree distribution 
fig = plt.figure()
nx.draw(star, node_size = 2000, node_color = 'yellow')
plt.show()

fig = plt.figure()
plt.hist(nx.degree(star).values())
plt.axis([0, 8, 0, 8])
plt.xlabel('Node Degree')
plt.ylabel('Frequency')
plt.show()

# create an adjacency matrix object for the star network
# use nodelist argument to order the rows and columns
star_mat = nx.adjacency_matrix(star,\
    nodelist = ['Amy', 'Bob', 'Dee', 'Joe', 'Lea', 'Max', 'Tia'])
print(star_mat)  # undirected networks are symmetric

# determine the total number of links for the star network (n-1)
print(np.sum(star_mat)/2)

# ---------------------------
# circle (undirected network)
# ---------------------------
# define graph object for undirected circle network
# using a list of links for pairs of nodes
circle = nx.Graph()
circle.add_edges_from([('Abe', 'Bea'), ('Abe', 'Rob'), ('Bea', 'Dag'),\
    ('Dag', 'Eve'), ('Eve', 'Jim'), ('Jim', 'Kat'), ('Kat', 'Rob')])
    
# examine the degree of each node
print(nx.degree(circle))
    
# plot the circle network and degree distribution 
fig = plt.figure()
nx.draw(circle, node_size = 2000, node_color = 'yellow')
plt.show()

fig = plt.figure()
plt.hist(nx.degree(circle).values())
plt.axis([0, 8, 0, 8])
plt.xlabel('Node Degree')
plt.ylabel('Frequency')
plt.show()

# create an adjacency matrix object for the circle network
# use nodelist argument to order the rows and columns
circle_mat = nx.adjacency_matrix(circle,\
    nodelist = ['Abe', 'Bea', 'Dag', 'Eve', 'Jim', 'Kat', 'Rob'])
print(circle_mat)  # undirected networks are symmetric

# determine the total number of links for the circle network 
print(np.sum(circle_mat)/2)

# -------------------------
# line (undirected network)
# -------------------------
# define graph object for undirected line network
# using a list of links for pairs of nodes
line = nx.Graph()
line.add_edges_from([('Ali', 'Ben'), ('Ali', 'Ela'), ('Ben', 'Ian'),\
    ('Ela', 'Mya'), ('Ian', 'Roy'), ('Mya', 'Zoe')])
    
# examine the degree of each node
print(nx.degree(line))

# plot the line network and degree distribution 
fig = plt.figure()
nx.draw(line, node_size = 2000, node_color = 'yellow')
plt.show()

fig = plt.figure()
plt.hist(nx.degree(line).values())
plt.axis([0, 8, 0, 8])
plt.xlabel('Node Degree')
plt.ylabel('Frequency')
plt.show()

# create an adjacency matrix object for the line network
# use nodelist argument to order the rows and columns
line_mat = nx.adjacency_matrix(line,\
    nodelist = ['Ali', 'Ben', 'Ela', 'Ian', 'Mya', 'Roy', 'Zoe'])
print(line_mat)  # undirected networks are symmetric

# determine the total number of links for the line network (n-1)
print(np.sum(line_mat)/2)

# ---------------------------
# clique (undirected network)
# ---------------------------
# define graph object for undirected clique
a_clique = nx.Graph()
a_clique.add_edges_from([('Ada', 'Ala'), ('Ada', 'Ami'), ('Ada', 'Ana'),\
('Ada', 'Ann'), ('Ada', 'Ara'), ('Ada', 'Ava'), ('Ala', 'Ami'),\
('Ala', 'Ana'), ('Ala', 'Ann'), ('Ala', 'Ara'), ('Ala', 'Ava'),\
('Ami', 'Ana'),('Ami', 'Ann'), ('Ami', 'Ara'), ('Ami', 'Ava'),\
('Ana', 'Ann'), ('Ana', 'Ara'), ('Ana', 'Ava'),\
('Ann', 'Ara'), ('Ann', 'Ava'), ('Ara', 'Ava')])

# examine the degree of each node
print(nx.degree(a_clique))

# plot the clique and degree distribution 
fig = plt.figure()
nx.draw_circular(a_clique, node_size = 2000, node_color = 'yellow')
plt.show()

fig = plt.figure()
plt.hist(nx.degree(a_clique).values())
plt.axis([0, 8, 0, 8])
plt.xlabel('Node Degree')
plt.ylabel('Frequency')
plt.show()

# create an adjacency matrix object for the line network
# use nodelist argument to order the rows and columns
a_clique_mat = nx.adjacency_matrix(a_clique,\
    nodelist = ['Ada', 'Ala', 'Ami', 'Ana', 'Ann', 'Ara', 'Ava'])
print(a_clique_mat)  # undirected networks are symmetric

# determine the total number of links for the clique n(n-1)/2
print(np.sum(a_clique_mat)/2)

# -------------------------------
# tree (directed network/digraph)
# -------------------------------
# define graph object for undirected tree network
# using a list of links for pairs of from-to nodes
tree = nx.DiGraph()
tree.add_edges_from([('Art', 'Bev'), ('Art', 'Dan'), ('Bev', 'Fay'),\
    ('Bev', 'Lee'), ('Dan', 'Mia'), ('Mia', 'Sal'), ('Mia', 'Van')])
    
# examine the degree of each node
print(nx.degree(tree))    
    
# create an adjacency matrix object for the line network
# use nodelist argument to order the rows and columns
tree_mat = nx.adjacency_matrix(tree,\
    nodelist = ['Art', 'Bev', 'Dan', 'Fay', 'Lee', 'Mia', 'Sal', 'Van'])
print(tree_mat)  # directed networks are not symmetric

# determine the total number of links for the tree 
# upper triangle only has values
print(np.sum(tree_mat))

# plot the degree distribution 
fig = plt.figure()
plt.hist(nx.degree(tree).values())
plt.axis([0, 8, 0, 8])
plt.xlabel('Node Degree')
plt.ylabel('Frequency')
plt.show()
    
# examine alternative layouts for plotting the tree 
# plot the network/graph with default layout 
fig = plt.figure()
nx.draw(tree, node_size = 2000, node_color = 'yellow')
plt.show()

# spring layout
fig = plt.figure()
nx.draw_spring(tree, node_size = 2000, node_color = 'yellow')
plt.show()

# circlular layout
fig = plt.figure()
nx.draw_circular(tree, node_size = 2000, node_color = 'yellow')
plt.show()

# concentric circles layout
fig = plt.figure()
nx.draw_shell(tree, node_size = 2000, node_color = 'yellow')
plt.show()

# note. plotting as tree may require pygraphviz

# Suggestions for the student: Define alternative network structures.    
# Use matplotlib to create their plots. Create the corresponding 
# adjacency matrices and compute network descriptive statistics,
# beginning with degree centrality. Plot the degree distribution
# for each network.  Read about pygraphviz and try to plot a tree.
